src/components/tl/cuda/alltoall/alltoall_ce.c

```diff
@@ -0,0 +1,181 @@
+                   exit_err, status, UCC_TL_TEAM_LIB(team));
+    __sync_synchronize();
+    asm volatile("": : :"memory");
+    sync->seq_num[0] = task->seq_num;
+    for (i = 0; i < team->size; i++) {
+        peer_sync = TASK_SYNC(task, i);
+        while (peer_sync->seq_num[0] != task->seq_num);
+    }
+
+    for (i = 0; i < team->size; i++) {
+        if (i == team->rank) {
+            continue;
+        }
+        peer_sync = TASK_SYNC(task, i);
+        cache = ucc_tl_cuda_get_cache(team, i);
+        if (ucc_unlikely(!cache)) {
+            status = UCC_ERR_NO_MESSAGE;
+            goto exit_err;
+        }
+        status = ucc_tl_cuda_map_memhandle(peer_sync->ptr, peer_sync->length,
+                                           peer_sync->handle,
+                                           &task->alltoall_ce.peer_map_addr[i],
+                                           cache);
+        if (UCC_OK != status) {
+            ucc_error("ucc_cuda_ipc_map_memhandle failed");
+            return UCC_ERR_INVALID_PARAM;
+        }
+    }
+    return UCC_OK;
+
+exit_err:
+    return status;
+}
+
+static ucc_status_t ucc_tl_cuda_alltoall_ce_post_copies(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_coll_args_t    *args = &TASK_ARGS(task);
+    ucc_tl_cuda_sync_t *sync = TASK_SYNC(task, team->rank);
+    ucc_tl_cuda_sync_t *peer_sync;
+    size_t send_len;
+    ucc_rank_t i, peer;
+    void *src, *dst;
+    ucc_status_t status;
+
+    send_len = (args->src.info.count / team->size) *
+               ucc_dt_size(args->src.info.datatype);
+    for (i = 0; i < team->size; i++) {
+        peer = (team->rank + i) % team->size;
+        peer_sync = TASK_SYNC(task, peer);
+        if (peer == team->rank) {
+            src = args->src.info.buffer;
+        } else {
+            src = PTR_OFFSET(task->alltoall_ce.peer_map_addr[peer],
+                             peer_sync->offset);
+            CUDACHECK_GOTO(cudaStreamWaitEvent(team->stream,
+                                               sync->data[peer].ipc_event_remote,
+                                               0),
+                           exit, status, UCC_TASK_LIB(task));
+        }
+        src = PTR_OFFSET(src, team->rank * send_len);
+        dst = PTR_OFFSET(args->dst.info.buffer, peer * send_len);
+        CUDACHECK_GOTO(cudaMemcpyAsync(dst, src, send_len, cudaMemcpyDeviceToDevice,
+                                       team->stream),
+                       exit, status, UCC_TASK_LIB(task));
+    }
+
+    status = ucc_mc_ee_event_post(team->stream, task->alltoall_ce.copy_done,
+                                  UCC_EE_CUDA_STREAM);
+exit:
+    return status;
+
+}
+
+ucc_status_t ucc_tl_cuda_alltoall_ce_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_status_t status;
+
+    status = ucc_tl_cuda_alltoall_setup(task);
```

Comment:
do we need to split alltoall_setup into 2 functions?
1st: local memory_setup - can be done during tl_cuda_alltoall_init
2nd: writing input dependency event + shm exchange - happens either in start or in early_triggered_post (using old terminology)

Reply:
done